src/main/java/edu/hm/hafner/analysis/parser/PnpmAuditParser.java

```diff
@@ -0,0 +1,124 @@
+
+    private static final long serialVersionUID = 4140706319863200922L;
+
+    @Override
+    protected void parseJsonObject(final Report report, final JSONObject jsonReport, final IssueBuilder issueBuilder) {
+        JSONObject results = jsonReport.optJSONObject("advisories");
+        if (results != null) {
+            parseResults(report, results, issueBuilder);
+        }
+    }
+
+    private void parseResults(final Report report, final JSONObject jsonReport, final IssueBuilder issueBuilder) {
+        for (String key : jsonReport.keySet()) {
+            JSONObject vulnerability = (JSONObject) jsonReport.get(key);
+
+            if (!vulnerability.isEmpty()) {
+                report.add(convertToIssue(vulnerability, issueBuilder));
+            }
+        }
+    }
+
+    private Issue convertToIssue(final JSONObject vulnerability, final IssueBuilder issueBuilder) {
+        return issueBuilder.setModuleName(vulnerability.optString("module_name", VALUE_NOT_SET))
+                .setCategory(formatCategory(vulnerability))
+                .setSeverity(mapSeverity(vulnerability.optString("severity", "UNKNOWN")))
+                .setType(mapType(vulnerability))
+                .setMessage(vulnerability.optString("overview", "UNKNOWN"))
+                .setDescription(formatDescription(vulnerability))
+                .buildAndClean();
+    }
+
+    private String mapType(final JSONObject vulnerability) {
+        final String cve = (String) vulnerability.optJSONArray("cves").opt(0);
+
+        return cve != null ? cve : VALUE_NOT_SET;
+    }
+
+    private String formatCategory(final JSONObject vulnerability) {
+        String moduleName = vulnerability.optString("module_name");
+        String title = vulnerability.optString("title");
+
+        if (moduleName != null && title != null) {
+            return title.replace(moduleName, "").replace(" in ", "");
+        }
+
+        return VALUE_NOT_SET;
+    }
+
+    @SuppressFBWarnings("IMPROPER_UNICODE")
+    private Severity mapSeverity(final String string) {
+        if (PNPM_VULNERABILITY_SEVERITY_INFO.equalsIgnoreCase(string)) {
+            return Severity.WARNING_LOW;
+        }
+        else if (PNPM_VULNERABILITY_SEVERITY_LOW.equalsIgnoreCase(string)) {
+            return Severity.WARNING_LOW;
+        }
+        else if (PNPM_VULNERABILITY_SEVERITY_MODERATE.equalsIgnoreCase(string)) {
+            return Severity.WARNING_NORMAL;
+        }
+        else if (PNPM_VULNERABILITY_SEVERITY_HIGH.equalsIgnoreCase(string)) {
+            return Severity.WARNING_HIGH;
+        }
+        else if (PNPM_VULNERABILITY_SEVERITY_CRITICAL.equalsIgnoreCase(string)) {
+            return Severity.ERROR;
+        }
+        else {
+            return Severity.ERROR;
+        }
+    }
+
+    private String formatDescription(final JSONObject vulnerability) {
+        final String moduleName = vulnerability.optString("module_name", VALUE_NOT_SET);
```

Comment:
Make sure that all properties are not `null` and use a meaningful default. Maybe it would make sense to remove the `div` if the entry is `null`? Then you can write a helper method?

Reply:
I refactored this section a little. The behaviour is now, that every value is checked for null and empty strings. Only if there is a value defined, the section is added to the description. 

I added some private helper methods to handle different scenarios. They may have the potential to be enhanced & extended, so that they are used as helper utils in many parsers. IMO, this should be handled in a following PR then 😊

Reply:
I recognized that my new helper functions `getValueAsContainerTag` (btw do you have ideas for better naming?) may return null, which is noted by spot bugs as a issue. In this case I have knowingly chosen this design.
Annotating it with `@Nullable` is also forbidden. Any suggestions or do we just keep it as is?

Reply:
The correct annotation is `@CheckForNull`!

Reply:
A more modern alternative would be to return an `Optional` and write something like:

`getValueAsContainerTag.ifPresent(list::add)`

Reply:
I refactored it to work with optionals, nice idea.